ENH: Significant speedup for ortho and special ortho group

JesseLivezey opened the pull request:
This PR makes 2 changes to `ortho_group.rvs()` and `special_ortho_group.rvs()` that significantly speed up the calculation.

1) The former `Hx` matrix is an identity minus a single outer product. In the next step it is matrix multiplied. It is faster to re-order this process as H.dot(I - xx^t) --> H - H.dot(x)x^T. In this re-ordering, you never have to do a matrix-matrix multiply.
2) The former `mat` matrix was block diagonal with the identity from 0 to n and `Hx` as the second block. On every loop, it was matrix multiplied by by `H`. The identity block means that you can just do the computation on a smaller and smaller subset of the columns of `H` at every iteration.

This changes the scaling of the computation and so it gives large speedups for `dim` > 100.

![ortho_group2](https://user-images.githubusercontent.com/2081702/56096560-b4a3fb00-5e9e-11e9-8b0e-630445ddc7e1.png)

![special_ortho2](https://user-images.githubusercontent.com/2081702/56096562-bc639f80-5e9e-11e9-8b45-6f5d358362b1.png)


@rgommers, this is a follow-up to #10058. It turns out I missed the most obvious ways to speed these function up in the first pass.

ilayn commented on scipy/stats/_multivariate.py:
```diff
@@ -3394,10 +3394,7 @@ def rvs(self, dim, size=1, random_state=None):
-            mat = np.eye(dim)
-            mat[n:, n:] = Hx
-            H = np.dot(H, mat)
+            H[:, n:] = H[:, n:] - np.outer(np.dot(H[:, n:], x), x)
```
Wouldn't you use `-=` here instead?

JesseLivezey replied:
Yep, I'll wait for the tests to finish and update.

ev-br commented on scipy/stats/_multivariate.py:
```diff
@@ -3394,10 +3394,7 @@ def rvs(self, dim, size=1, random_state=None):
-            mat = np.eye(dim)
-            mat[n:, n:] = Hx
-            H = np.dot(H, mat)
+            H[:, n:] -= np.outer(np.dot(H[:, n:], x), x)
```
Does it need `np.outer`? It's scalar-vector product, is it not?

JesseLivezey replied:
Unless I'm missing another simplification, I think this does need outer. The result of `np.dot(H[:, n:], x)` is a vector and you want the outer product of that vector and `x`.

JesseLivezey commented:
Some added detail for potential future readers:

Previously, there was a `(dim x dim) x (dim x dim)` matrix multiply in a loop of size `dim`. The complexity of this is O(`dim^4`) assuming square matrix multiplies are O(`dim^3`).

With this change, there is a `(dim x n) x n` matrix-vector product (O(`dim x n`)) followed by a `n x n` vector-vector outer product (O(`n x n`)) where `n` is the loop index inside a loop of size `dim`. Overall, this should be O(`dim^3)`.

These changes just re-order the previous calculations and skip some trivial computations (multiplication by identity, etc.). Up to small numerical differences, the results of this should be identical to the previous version.

I haven't tried this yet, but these changes might also make it possible to vectorize over the `size` argument rather than looping over it (and turn matrix-vector back into matrix-matrix operations).

Unrelated to this PR, I've rewritten this code in pytorch to run on GPUs. I need to generate 1000s of samples with 1000s of dimensions and see pretty good (~25x) speedups for large dims.

https://github.com/JesseLivezey/pytorch_group_sampler

JesseLivezey commented:
@tylerjereddy Sadly, I am to blame since I added that (broken) benchmark in a recent PR. :) I pushed the fix.

Looking at the scipy `asv`, I noticed that another unrelated benchmark I added a while ago doesn't have any data
https://pv.github.io/scipy-bench/#spatial.Radius.time_query_ball_point_nosort
which is here
https://github.com/scipy/scipy/blob/8532ca64d5a76e2a014ea1787cd72b79087a5696/benchmarks/benchmarks/spatial.py#L130

Is it missing something obvious? I've never actually used `asv`.

tylerjereddy commented:
If you could add a small entry summarizing the improvements from this PR at the release note page that will help me out: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0

Glancing at the other benchmark you mention, I don't see an immediately obvious problem. You could always try running it with a command similar to what I've used above.

JesseLivezey commented:
@tylerjereddy I added a comment. Thanks!
